src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -185,26 +181,27 @@ private unsafe Task<int> ReadAsyncInternal(Memory<byte> destination, Cancellatio
             if (r == -1)
             {
                 // For pipes, when they hit EOF, they will come here.
-                if (errorCode == ERROR_BROKEN_PIPE)
+                if (errorCode == Interop.Errors.ERROR_BROKEN_PIPE)
                 {
                     // Not an error, but EOF.  AsyncFSCallback will NOT be
                     // called.  Call the user callback here.
 
                     // We clear the overlapped status bit for this special case.
                     // Failure to do so looks like we are freeing a pending overlapped later.
                     intOverlapped->InternalLow = IntPtr.Zero;
-                    completionSource.SetCompletedSynchronously(0);
+                    valueTaskSource.ReleaseNativeResource();
+                    return new ValueTask<int>(valueTaskSource.NumBufferedBytes);
```

Comment:
Do you know why this wasn't performed before and does it only bring benefit to this special case? 

It seems to me that NumBufferedBytes is always zero, could we return `new ValueTask(0)` here instead since that's more clear?

Reply:
I see what you are referring to. I was trying to mimic what we had in `FileStreamCompletionSource` in the [`SetCompletedAsynchronously` method](https://github.com/dotnet/runtime/blob/a913740aac5353212a9b95283ae38b9964f18aa5/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamCompletionSource.Win32.cs#L69-L73), but without having to add that method. I can probably also remove `NumBufferedBytes` property if I'm just going to pass `0`.

Reply:
NumBufferedBytes was only used in the buffering logic of ReadAsync of legacy code, since the new code doesn't do that and since FileStreamValueTaskSource is only used in new code, I think you can safely remove it.